sdk/storage/AzureStorageBlob/Source/DataStructures/URLSessionTransferManager+ResumableOperationQueueDelegate.swift

```diff
@@ -35,19 +35,36 @@ extension URLSessionTransferManager: ResumableOperationQueueDelegate {
             let blocks = blobTransfer.totalBlocks - blobTransfer.incompleteBlocks
             return BlobUploadProgress(bytes: Int(blocks), totalBytes: Int(blobTransfer.totalBlocks))
         case .download:
-            let blocks = blobTransfer.totalBlocks - blobTransfer.incompleteBlocks
-            return BlobDownloadProgress(bytes: Int(blocks), totalBytes: Int(blobTransfer.totalBlocks))
+            guard let downloader = blobTransfer.downloader else { return nil }
+            return BlobDownloadProgress(bytes: downloader.progress, totalBytes: downloader.fileSize)
         default:
             return nil
         }
     }
 
     public func operation(_ operation: ResumableOperation?, didChangeState state: TransferState) {
         if let transfer = (operation as? ResumableTransfer)?.transfer {
-            transfer.state = state
             saveContext()
             let progress = buildProgressInfo(forTransfer: transfer)
-            delegate?.transferManager(self, didUpdateTransfer: transfer, withState: state, andProgress: progress)
+            switch state {
+            case .failed:
+                // Block transfers should propagate up to their BlobTransfer and only notify that the
+                // entire Blob transfer failed.
+                if let transferError = (transfer as? BlobTransfer)?.error as? NSError {
+                    if [-1009, -1005].contains(transferError.code) {
+                        pause(transfer: transfer)
+                    } else {
+                        delegate?.transferManager(self, didFailTransfer: transfer, withError: transferError)
+                    }
+                } else {
+                    // ignore BlockTransfer failures
```

Comment:
I know it's not implemented yet but we should revisit this when MultiBlob transfers are in.

Reply:
I'm actually thinking it should notify the delegate that the block transfer failed. I was previously notifying just the parent when blocks completed, but had to switch notifying both. The idea being that the app developer can decide which perspective they care about. So I am open to bubbling this failure up. Is that what you were thinking?